### DISCUSSION TITLE
next dev - HMR is giving me grief (multiple reloads, refresh looping on errors)

### DISCUSSION THREAD

**Discussion by elsigh** [vercel/next.js#12648] [Help] [answered]
While developing locally I'm encountering two situations that are undesirable and make me want to disable HMR altogether, but I wonder if I'm just doing something wrong, so looking for advice here.

If I make 3 independent edits and click save in my editor and come back to my app it reloads three times (and fires off three network requests since I'm fetching data immediately in an effect. Is that sort of time-travel reload intentional? Apologies if this is just how HMR is supposed to work and I'm asking here on the NextJS forum ;) But still, maybe you've got some advice?

More challenging is that when I create an error (quite often a ReferenceError to an undefined variable while refactoring or fat fingers) I'll come over to my browser and it goes into a death refresh loop, just reloading over and over and over (and firing network requests ala above). I usually can't keep up with the console logs and then just kill the Chrome tab in the Task Manager because my laptop is starting to get overworked. In that case I lose the error as well, so it's a pretty rough developer experience. Thoughts?

Thank you!

**Comment by marcusmota** [comment]
Same issue here, sometimes I need to close the tab and open another one due too many HMR / requests being fired at once.

**Reply by Timer** [reply]
Please open a new bug report with any HMR behavior you believe is weird or incorrect. Also be sure you're on Next.js 9.4 or newer.

**Comment by Timer** [comment, marked as answer]
<details>
<summary>Old suggestion</summary>
Can you please try React Refresh on `next@canary`? You can enable it in `next.config.js`:
```js
// next.config.js
module.exports = {
  experimental: {
    reactRefresh: true,
  },
}
```
</details>

Next.js 9.4 has a completely new HMR system that is more reliable and correct! Please upgrade. 😄

**Reply by elsigh** [reply]
For sure! Will report back, thanks for a suggestion.

**Reply by Timer** [reply]
This is enabled by default on `next@canary`, going out today in 9.4.

**Reply by elsigh** [reply]
Not sure how it's supposed to work, but it seems it turns off hmr.. e.g. I save in my editor, the page gets recompiled but nothing happens in my browser. I don't mind refreshing honestly, and thought just figuring out how to turn off hmr would be a solution of sorts, but not sure what _should_ be happening.

**Reply by Timer** [reply]
Have you named all of your components?

HMR will not work if you do the following:
```js
export default () => {
  return <div>...</div>
}
```

**Reply by elsigh** [reply]
My components are all named and hmr was working prior to (and causing me trouble ;)

**Reply by elsigh** [reply]
Interestingly though I just created a reference error and instead of spinning out of control I got a nice popup ui pointing to the error by line!

**Reply by elsigh** [reply]
actually it may just be working after all for me now.. no flash of UI either!

**Reply by Timer** [reply]
Great! I'll mark this as the answer then.

**Reply by Timer** [reply]
Please open a new bug report with any HMR behavior you believe is weird or incorrect.
